strawberryfields/tdm/tdmprogram.py

```diff
@@ -81,69 +85,176 @@ def validate_measurements(circuit, N):
+        MeasureHomodyne(0) | (q[1])  # shot 1, timebin 1, spatial mode 2  (sample 7)
+
+    would return the dictionary
+
+    .. code-block:: pycon
+
+        {
+            0: np.array([(sample 0), (sample 2), (sample 4), (sample 6)]),
+            2: np.array([(sample 1), (sample 5)]),
+            1: np.array([(sample 3), (sample 7)]),
+        }
+
+    which would then be reshaped, and returned, as follows:
+
+    .. code-block:: pycon
+
+        {
+            0: np.array([[(sample 0), (sample 2)],
+                         [(sample 4), (sample 6)]]),
+            2: np.array([[(sample 1), (sample 3)],
+                         [(sample 5), (sample 7)]]),
+        }
+
     Args:
         all_samples (dict[int, list]): the raw measured samples
         modes (Sequence[int]): the modes that are measured in the circuit
         N (Sequence[int]): the number of concurrent modes per belt/spatial modes
+        timebins (int): the number of timebins/temporal modes in the program per shot
 
     Returns:
-        new_samples (dict[int, list]): the re-shaped samples
+        dict[int, array]: the re-shaped samples, where each key correspond to a spatial
+            mode and the values have shape ``(shots, timebins)``
     """
+    # calculate the total number of samples and the order in which they were measured
     num_of_values = len([i for j in all_samples.values() for i in j])
-    mode_order = _get_mode_order(num_of_values, N)
+    mode_order = _get_mode_order(num_of_values, modes, N, timebins)
 
-    # go backwards through all_samples and add them into the correct mode
+    # iterate backwards through all_samples and add them into the correct mode
     new_samples = dict()
-    for i, m in enumerate(mode_order):
-        idx = modes[i % len(N)]
-        if idx not in new_samples:
-            new_samples[idx] = []
-        new_samples[idx].append(all_samples[m].pop(0))
-    return new_samples
+    timebin_idx = 0
+    for i, mode in enumerate(mode_order):
+        mode_idx = modes[i % len(N)]
+
+        if mode_idx not in new_samples:
+            # create an entry for the new mode with a nested list for each timebin
+            new_samples[mode_idx] = [[] for _ in range(timebins)]
+
+        sample = all_samples[mode].pop(0)[0]
+        new_samples[mode_idx][timebin_idx].append(sample)
+
+        # populate each spatial mode in one timebin before moving on to the next timebin
+        # when each timebin has been filled, move to the next shot
+        last_mode_in_timebin = (i + 1) % len(N) == 0
+        if last_mode_in_timebin:
+            timebin_idx = (timebin_idx + 1) % timebins
+
+    # transpose each value so that it has shape `(shots, timebins)`
+    return {k: np.array(v).T for k, v in new_samples.items()}
+
+
+def move_vac_modes(samples, N, crop=False):
+    """Post-processing function for TDM samples. Moves all measured vacuum modes
+    from the first shot of the samples array to the end of the last shot.
```

Comment:
"Post-processing function for TDM samples" is a bit superfluous, we can probably remove it:
```suggestion
    """Moves all measured vacuum modes from the first shot of the
    returned TDM samples array to the end of the last shot.
```

Reply:
I have to admit, I still don't quite follow how come this function was needed. It would be good to get some context here!